fs-uae - added integration into emulationstation. Supporting ZIPs containing multiple disks

verybadsoldier opened the pull request:
So, me again ;)

Another try on integrating fs-uae into ES. Now taking the ZIP functions into account (so no multi-disk-discovery needed anymore).

Some explanations on some points (does not make them necessarily good decisions ofc):

-  I thought about adding Amiga to platforms.cfg and to store the supported extensions there. But the extensions are depending on the emulator. So fs-uae supports IPF, ADF etc. (and now ZIP) while uae4all seems to be ADF only. I don't know if platforms.cfg is supposed to handle this so I added the extensions as a variable to the emulator script. Actually I wanted to have it (FSUAE_EXTS) readonly but when I declare it readonly it is always empty later in the configure function. I could not find out why.

- I changed the config directory symbolic link from `~/.config/fs-uae` to `~/Documents/FS-UAE/Configuration` since this is the directory that fs-uae creates on first run to store and read the config from.

- I added a hopefully sensible default config for fs-uae to have it run in fullscreen by default. Also some other parameters are in there so they can be changed easily by the user.

Looking forward to your input!

joolswills commented on scriptmodules/emulators/fs-uae.sh:
```diff
@@ -9,9 +9,11 @@
 # at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
 #
 
+FSUAE_EXTS=(.adf .ipf .dms .adz .zip)
```
no globals here please.

joolswills commented on scriptmodules/emulators/fs-uae.sh:
```diff
@@ -9,9 +9,11 @@
 rp_module_id="fs-uae"
 rp_module_desc="Amiga emulator - FS-UAE integrates the most accurate Amiga emulation code available from WinUAE"
-rp_module_help="ROM Extension: .adf"
+rp_module_help="ROM Extension: ${FSUAE_EXTS[*]}\n\nCopy your Amiga games to $romdir/amiga\n\nCopy a required BIOS file (e.g. kick13.rom) to $biosdir"
```
I think I would prefer just putting the extensions in platforms.cfg - the emulators themselves document what they support and we have this in other cases where some emulators don't support all formats.

joolswills commented on scriptmodules/emulators/fs-uae.sh:
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
 
-    mkUserDir "$home/.config"
+    # copy configuring start script
+    mkdir "$md_inst/bin/"
```
minor, but no need for /

joolswills commented on scriptmodules/emulators/fs-uae.sh:
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
-    mkUserDir "$home/.config"
+    # copy configuring start script
+    mkdir "$md_inst/bin/"
+    cp "$scriptdir/scriptmodules/$md_type/$md_id/fs-uae.sh" "$md_inst/bin/"
```
use `$md_data`

joolswills commented on scriptmodules/emulators/fs-uae.sh:
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
-    chmod a+x "$romdir/amiga/+Start FS-UAE.sh"
-    chown $user:$user "$romdir/amiga/+Start FS-UAE.sh"
+    # copy default config file
+    cp -v "$scriptdir/scriptmodules/$md_type/$md_id/Default.fs-uae" "$md_conf_root/amiga"
```
tiny config - can use iniSet - also make sure we don't overwrite existing configs on upgrade. (configure should not overwrite existing configs).

joolswills commented on scriptmodules/emulators/fs-uae/fs-uae.sh:
```diff
@@ -0,0 +1,49 @@
+config="$configdir/amiga/Default.fs-uae"
+
+user="$SUDO_USER"
+[[ -z "$user" ]] && user="$(id -un)"
```
no need for this - the script is not run under sudo.

joolswills commented on scriptmodules/emulators/fs-uae/fs-uae.sh:
```diff
@@ -0,0 +1,49 @@
+
+user="$SUDO_USER"
+[[ -z "$user" ]] && user="$(id -un)"
+home="$(eval echo ~$user)"
```
can just use `$HOME` in this case

joolswills commented:
>I changed the config directory symbolic link from ~/.config/fs-uae to ~/Documents/FS-UAE/Configuration since this is the directory that fs-uae creates on first run to store and read the config from.

are you saying that `~/.config/fs-uae` is not used or created by fs-uae ?

joolswills commented:
thanks - made a few comments above.

verybadsoldier commented:
Ok, thanks for your hints, next iteration is here :)

Regarding config path:
> are you saying that ~/.config/fs-uae is not used or created by fs-uae ?

fs-uae reads from `Documents/FS-UAE/Configuration` but also from `~/.config/fs-uae`. The latter is the "Hidden Default Configuration File" as documented here: https://fs-uae.net/configuration-files
As I understand it, the main purpose of the hidden config is to use the parameter `base_dir` to redirect the whole fs-uae path to something custom.
So I think using `Documents` is the more straightforward way but I can also change it to `.config` if you prefer it.

I am still not really happy with the handling of the extensions. I have added them to platforms.cfg but also entered them in `rp_module_help`. Also the start script needs them at runtime. So I pass the extensions as parameter to the start script (read from platforms.cfg), because I think the start script cannot access platforms.cfg (and because platforms.cfg is not specific for the current emulator).
(In general I think it is a bit dirty for emulator scripts to use the extensions information from platforms.cfg because they are not specific for the emulator and also might change later when new emulators are added.)
I can live with it but if you have ideas how to improve it I would be happy to implement them.

Also when invoking fs-uae I removed the parameter which passes a specific configfile so I rely on fs-uae finding the default config (Default.fs-uae) on its own, which should be fine (hopefully). It simplified some lines of code in the start script.

joolswills commented:
I'm happy with the extensions being in platforms.cfg.

Instead of passing by argument, you could include them directly in the launch script - by adding them in the same way we add extensions in the asplashscreen init.d script as it is installed. or put the script in as a heredoc as with vice, and put the extensions in the script there.

joolswills commented:
Anyway, it's also fine as it is now, so I will merge - I may tweak a few things later.

verybadsoldier commented:
Ok, great, again thanks!
